include/nil/crypto3/math/multithreading/parallelization_utils.hpp

```diff
@@ -0,0 +1,139 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_PARALLELIZATION_UTILS_HPP
+#define CRYPTO3_PARALLELIZATION_UTILS_HPP
+
+#include <future>
+
+#include <nil/crypto3/math/multithreading/thread_pool.hpp>
+
+namespace nil {
+    namespace crypto3 {
+
+        template<class ReturnType>
+        void wait_for_all(const std::vector<std::future<ReturnType>>& futures) {
+            for (auto& f: futures) {
+                f.wait();
+            }
+        }
+
+        // Similar to std::transform, but in parallel. We return void here for better usability for our use cases.
+        template<class InputIt1, class InputIt2, class OutputIt, class BinaryOperation>
+        void parallel_transform(InputIt1 first1, InputIt1 last1, InputIt2 first2,
+                                OutputIt d_first, BinaryOperation binary_op,
+                                ThreadPool::PoolID pool_id = ThreadPool::PoolID::LOW_LEVEL_POOL_ID) {
+
+            wait_for_all(ThreadPool::get_instance(pool_id).block_execution<void>(
+                std::distance(first1, last1),
+                // We need the lambda to be mutable, to be able to modify iterators captured by value.
+                [first1, last1, first2, d_first, binary_op](std::size_t begin, std::size_t end) mutable {
+                    std::advance(first1, begin);
+                    std::advance(first2, begin);
+                    std::advance(d_first, begin);
+                    for (std::size_t i = begin; i < end && first1 != last1; i++) {
+                        *d_first = binary_op(*first1, *first2);
+                        ++first1;
+                        ++first2;
+                        ++d_first;
+                    }
+                }));
+        }
+
+        // Similar to std::transform, but in parallel. We return void here for better usability for our use cases.
+        template< class InputIt, class OutputIt, class UnaryOperation >
+        void parallel_transform(InputIt first1, InputIt last1,
+                                OutputIt d_first, UnaryOperation unary_op,
+                                ThreadPool::PoolID pool_id = ThreadPool::PoolID::LOW_LEVEL_POOL_ID) {
```

Comment:
Nope, let's keep this defaults. Almost always we want the lower pool when doing transforms, and in case someone makes a wrong choice, he will immediately get a deadlock and notice that something's wrong anyway.